modules/sw360/sw360-client/src/main/java/org/eclipse/sw360/antenna/sw360/client/rest/SW360Client.java

```diff
@@ -0,0 +1,263 @@
+     * may have expired. The method checks whether the request failed with the
+     * special {@code RequestUnauthorizedException} exception.
+     *
+     * @param exception the exception the request failed with if any
+     * @return a flag whether a retry should be attempted
+     */
+    private boolean checkIfRetry(Throwable exception) {
+        if (exception instanceof UnauthorizedRequestException) {
+            getTokenProvider().invalidate(((UnauthorizedRequestException) exception).token);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Checks whether the given exception indicates an unauthorized request.
+     * This could mean that the access token has expired, and the request may
+     * be retried with a fresh token.
+     *
+     * @param exception the exception causing the request to fail
+     * @return true if the request was unauthorized, false otherwise
+     */
+    private static boolean isRequestUnauthorized(Throwable exception) {
+        return FutureUtils.isFailedRequestWithStatus(exception, HttpConstants.STATUS_ERR_UNAUTHORIZED);
+    }
+
+    /**
+     * Adds some checks to the original {@code ResponseProcessor} for a request
+     * to be executed. Before this processor is invoked, it is first checked
+     * whether the response has a successful status code. The status code
+     * indicating an unauthorized request is handled in a special way as it
+     * might cause the request to be retried.
+     *
+     * @param processor   the original {@code ResponseProcessor}
+     * @param tag         a tag to identify the request
+     * @param accessToken the current access token
+     * @param <T>         the type of the result
+     * @return the decorated {@code ResponseProcessor}
+     */
+    private static <T> ResponseProcessor<T> applyChecks(ResponseProcessor<T> processor, String tag,
+                                                        AccessToken accessToken) {
+        return checkUnauthorized(HttpUtils.checkResponse(processor, tag), tag, accessToken);
+    }
+
+    /**
+     * Returns a special {@code ResponseProcessor} that wraps another processor
+     * and checks whether the request failed because it was unauthorized. In
+     * this case, the exception is transformed and enriched with the (probably
+     * expired) access token.
+     *
+     * @param processor the {@code ResponseProcessor} to be wrapped
+     * @param token     the current access token
+     * @param <T>       the type of the result generated
+     * @return the result of the {@code ResponseProcessor}
+     */
+    private static <T> ResponseProcessor<T> checkUnauthorized(ResponseProcessor<T> processor, String tag,
+                                                              AccessToken token) {
+        return response -> {
+            try {
+                return processor.process(response);
+            } catch (FailedRequestException e) {
```

Comment:
Really necessary, does this specialized exception bring any value?

Reply:
see my other comment

Reply:
I still do not know whether this accesstoken thing is needed. This is a lot of complexity to transfer the access token through the call stack so that the invalidate side effect is security protected. Why does invalid need this in the first place, because an superfluous invalidate causes a new token to be aquired. So if called slightly to often this creates too much token renewals but so what?

Reply:
Sorry, when I see a race condition, I feel the urge to fix it.

The problem will occur in practice if there are many concurrent requests. Without special handling, for each of them the invalidation is done, and a new token is fetched. When this level of concurrency is used performance is certainly an issue. It would be a pity if the library then performs bad.

Reply:
Found an alternative solution. The access tokens are now fully managed by the AccessTokenProvider class. It detects itself when a token is expired; so the code of SW360Client could be simplified. The access token no longer has to be passed to the chain of futures.